Add mean to std.numeric

JackStouffer opened the pull request:
Now that https://github.com/D-Programming-Language/phobos/pull/2991 has been closed, I am reintroducing this function. I believe std.numeric to be a better fit than std.algorithm.iteration, so I moved it.

See https://github.com/D-Programming-Language/phobos/pull/3679 for previous discussion

Also fixes issue 14034

9il commented:
#2991 has been closed because it will be a part of the future `std.las` (linear algebra subroutines). `std.las` will have `mean` (Package Change) with different summation algorithms (API Change) and optimizations (Implementation Change). I will write `std.las` schedule, plan and first docs, so any one will be able to participate in this project.

CyberShadow commented:
@JackStouffer Since `std.las` never materialized, what do you think of reopening this?

`Average` is the only LINQ feature on https://github.com/wilzbach/linq that has no D equivalent.

JackStouffer commented:
@CyberShadow heh, blast from the past.

Sure, if people are interested, I'll resurrect this.

CyberShadow commented:
There is a `private:` in std.numeric on line 3332, you will need to override it or add the code above it.

JackStouffer commented:
@CyberShadow Fixed

CyberShadow commented:
```d
	int[] arr = null;
	writeln(mean(arr, 5));
```

This prints 0.

What does the seed do, anyway? I think it could use a bit more elaboration.

JackStouffer commented:
Yeah, the seed is mainly there to support user defined types. It doesn't really make sense when you're just using plain number types.

I'll add a note in the docs.

CyberShadow commented:
OK.

Shouldn't `mean` of an empty list return `nan` instead of 0 though?

JackStouffer commented:
@CyberShadow It will if you don't explicitly set the seed. Or, are you saying that there should be some `static if` special casing in the second overload?

CyberShadow commented:
Maybe I just don't understand the docs.

> Finds the mean (colloquially known as the average) of a range. If `r` does not provide the `length` member, then this function will do element by element summation, rather than the more accurate methods provided by $(REF sum, std,algorithm,iteration).

So far so good.

> An optional parameter `seed` may be passed to initially populate the summation.

Populate in what way? What effect does it have, exactly?

> This function will return `real.nan` if the range is empty and if the element type of `r` is a built-in numerical type.

The above conditions are both satisfied for my example above. So it should return `real.nan` instead of 0, no?

> Or, if the seed value is defined and not a built-in numerical type, then `ElementType!R.init` will be returned. 

A seed is given but it is a numerical type, so this shouldn't apply to my example above.

Also, is the seed *value* used anywhere, or just its type?

CyberShadow commented:
> Or, are you saying that there should be some `static if` special casing in the second overload?

Yep, I think nonsensical instantiations should be rejected outright to avoid confusion.

JackStouffer commented:
>Also, is the seed value used anywhere, or just its type?

The value is used in both the call to `sum` and in the `reduce` call.

>Yep, I think nonsensical instantiations should be rejected outright to avoid confusion.

Ok, changed the template constraints on the second overload and changed the docs.

CyberShadow commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+/// ditto
+auto mean(R, T)(R r, T seed)
+    if (isInputRange!R &&
+        (is(T == real) || !isNumeric!(T)) &&
```
Why allow `real` seeds?

JackStouffer replied:
to make the first overload compile

JackStouffer replied:
plus it also correctly returns nan if you do

```d
int[] a = [];
mean(a, real(0));
```

wilzbach replied:
Why you restrict the seed type at all?
Isn't `assert((cast(int[]) null).mean(0) == 0)` a valid use case?

JackStouffer replied:
>Isn't assert((cast(int[]) null).mean(0) == 0) a valid use case?

No. Taking the mean of an array and not giving back a FP value makes no sense from a statistical perspective. 

This is why this function is in std.numeric and not std.algorithm, because it's opinionated and not generic by design.

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), seed), r);
+
+        return pair[1] / pair[0];
```
Please let's use a more efficient algorithm here. Or to quote [Jonis alonen](http://jonisalonen.com/2013/deriving-welfords-method-for-computing-variance/)

> Do not use this method to compute variance __ever__. This is one of those cases where a mathematically simple approach turns out to give wrong results for being numerically unstable. In simple cases the algorithm will seem to work fine, but eventually you will find a dataset that exposes the problem with the algorithm.

To put this in code. The naive approach vs. an online.combination after Welford. I used [StatReport](https://github.com/libmir/mir/pull/90/files) here, it's a OutputRange for mean + variance, but[Summary](https://github.com/DlangScience/dstats/blob/master/source/dstats/summary.d#L839) of dstats is nice as well.

```d
auto arr = [10e19 + 4, 7, 10e9 + 13, 10e15 + 16];
writefln("Naive: %f", arr.mean);
writefln("Online: %f", arr.stat.mean);
```

yields:

```
Naive: 25002500002500001792.000000
Online: 25002500002500000008.000000
```

For reference, it should be [25002500002500000010](http://www.wolframalpha.com/input/?i=mean+%7B10e19+%2B+4,+7,+10e9+%2B+13,+10e15+%2B+16%7D).
Or [if you prefer to read a paper](http://people.ds.cam.ac.uk/fanf2/hermes/doc/antiforgery/stats.pdf) (the simple trick is equation 4 in chapter 1):

![image](https://user-images.githubusercontent.com/4370550/27997973-b72b0e36-6503-11e7-8018-4b698c4006c9.png)

This method is more precise, but not as efficient due to division in the loop. However, I argue that most users won't care and correctness is more important than performance for a standard library.

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+    assert(bigint_arr.mean(BigInt(0)) == BigInt("3"));
+    assert(bigint_arr2.mean(BigInt(0)) == BigInt("3"));
+}
+
```
A test for floating point accuracy would be nice.

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
```
Are you aware that `real` FP (in this summation) is _quite_ slower than `double`?

```d
/**
Compile with:
> ldc -release -O3 -mcpu=native -boundscheck=off
*/
import std.algorithm, std.range;

private void doNotOptimizeAway(T)(auto ref T t)
{
    import core.thread : getpid;
    import std.stdio : writeln;
    if(getpid() == 1) {
        writeln(*cast(char*)&t);
    }
}

/// ditto
auto mean(R, T)(R r, T seed)
{
    import std.algorithm.iteration : sum, reduce;
    static if (hasLength!R)
    {
        if (r.length == 1)
            return r.front;

        return sum(r, seed) / r.length;
    }
    else
    {
        import std.typecons : tuple;

        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
            (tuple(size_t(0), seed), r);

        return pair[1] / pair[0];
    }
}

void main() {
    import std.datetime;
    import std.stdio;
    import std.array;
    import std.conv;
    import std.random;
    auto arr = iota(100_000).array;
    auto bench = benchmark!(
        { doNotOptimizeAway({
            return arr.mean(real(0));
        }());},
        { doNotOptimizeAway({
            return arr.mean(double(0));
        }());},
        { doNotOptimizeAway({
            return arr.filter!(a => a >= 0).mean(real(0));
        }());},
        { doNotOptimizeAway({
            return arr.filter!(a => a >= 0).mean(double(0));
        }());},
    )(50_000);
    string[] names = ["mean (real)", "mean (double)", "mean (real, InputRange)", "mean (double, InputRange)"];

    foreach(j,r;bench)
        writefln("%-15s = %s", names[j], r.to!Duration);
}
```

```
> ldc -release -O3 -mcpu=native -boundscheck=off test.d && ./test
mean (real)     = 5 secs, 185 ms, 845 μs, and 3 hnsecs
mean (double)   = 3 secs, 869 ms, 445 μs, and 9 hnsecs
mean (real, InputRange) = 6 secs, 109 ms, 83 μs, and 8 hnsecs
mean (double, InputRange) = 5 secs, 940 ms, 972 μs, and 1 hnsec
```

JackStouffer replied:
I assume that if you want statistical values, accuracy will be more valuable than speed.

wilzbach replied:
> I assume that if you want statistical values, accuracy will be more valuable than speed.

FWIW with the test provided below `double` yields exactly the same inaccuracy (`25002500002500001792.000000`).
The point was that always picking `real` is a severe performance penalty with almost no neglectable accuracy improvements, so the user should at least have the choice.
Also there's another, very valid point about allowing `double`: it's consistent on all platforms. I purposefully ignored `real` on my Tinflex random project last year because the algorithm was very sensitive to small change ([more details about FP behavior with dmd](http://forum.dlang.org/post/hjaiavlfkoamenidomsa@forum.dlang.org) or as a [handy list of FP issues to keep in mind](https://github.com/libmir/mir/wiki/Floating-point-issues)).

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+        BigInt("1"), BigInt("2"), BigInt("3"), BigInt("6")
+    ]);
+    assert(bigint_arr.mean(BigInt(0)) == BigInt("3"));
+    assert(bigint_arr2.mean(BigInt(0)) == BigInt("3"));
```
Could we test `mean` on an empty array of a  user-defined typed (e.g. BigInt). Thanks!

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+    }
+
+    static if (hasLength!R)
+    {
```
As mentioned below, it's more precise to do division on every step, but of course a bit slower.

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+            return r.front;
+        }
+
+        return sum(r, seed) / r.length;
```
> @JackStouffer Since std.las never materialized

FYI there's `mir.math.sum`: https://github.com/libmir/mir-algorithm/blob/master/source/mir/math/sum.d
It's well tested and highly optimized.

JackStouffer replied:
If anyone is looking to port that to Phobos, I'd be willing to use it :P

JackStouffer commented on std/numeric.d:
```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
+
+    if (r.empty)
+    {
+        return Unqual!(ElementType!R).init;
```
The second one won't compile.

JackStouffer replied:
The problem as mentioned above is that mean should return a FP result when ever possible.

JackStouffer commented:
@wilzbach I tried your algorithm and I'm still getting the same result for the test case. What do I need to change?

```d
auto mean(R, T)(R r, T seed)
{
    T meanRes = seed;
    size_t i = 0;

    foreach (e; r) {
        immutable delta = (e - meanRes);
        meanRes += delta / ++i;
    }

    return meanRes;
}

auto arr = [10e19 + 4, 7, 10e9 + 13, 10e15 + 16];
writefln("Online: %f", arr.mean); // 25002500002500001792.000000
```

wilzbach commented:
> @wilzbach I tried your algorithm and I'm still getting the same result for the test case. What do I need to change?

Looks like your `seed` is `double`. Try this:

```d
void main()
{
    import std.stdio;
    auto arr = [10e19 + 4, 7, 10e9 + 13, 10e15 + 16];
    writefln("Online: %f", arr.mean(double(0)));
    writefln("Online: %f", arr.mean(real(0)));
}
```

[Run it online](https://is.gd/OOqPci)

JackStouffer commented:
Gah! Seems like this needs to be really re-worked.

JackStouffer commented:
@wilzbach I think addressed most of the issues.

wilzbach commented:
Behavior for empty input
--------------------------------

I thought a bit and returning `NaN` is definitely reasonable and what literally all other statistical languages do:

Language | Command | Results
------------- |----------------|----------|
Python |  `np.mean(list())` | `nan`
Julia | `mean(Array{Float64, 1}())` or `mean(Array{Int64, 1}())`   | `NaN`
R | `mean(double())` or `mean(int())` | `NaN`
Matlab / Octave |  ` mean([])` | `NaN`

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,117 @@ C swapRealImag(C)(C input)
+    The mean of `r` when `r` is non-empty.
+*/
+T mean(T = real, R)(R r)
+    if (isInputRange!R &&
```
Nit: The `if` constraint should be on the same indent level as the declaration.
See also: it's [part of the DStyle](https://github.com/dlang/dlang.org/pull/1534) and will be automated [soon](https://github.com/dlang-community/D-Scanner/pull/450).

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,117 @@ C swapRealImag(C)(C input)
+    BigInt[] bigint_arr3 = [];
+    assert(bigint_arr3.mean(BigInt(0)) == BigInt(0));
+}
+
```
One last idea: testing the behavior with subtypes:

```d
struct MyFancyDouble
{
   double v;
   alias v this;
}
```

(they currently use the second overload, however, intuitively as a user I would expect the first one).

JackStouffer replied:
>(they currently use the second overload, however, intuitively as a user I would expect the first one).

Yeah, well that's the problem with `alias this` and templates. `alias this` works great with function overloads, but with templates you have to explicitly cast a la the hacks in `std.file`

wilzbach commented on std/numeric.d:
```diff
@@ -3329,6 +3329,117 @@ C swapRealImag(C)(C input)
+    // inaccurate for integer division, which the user defined
+    // types might be representing
+    auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+        (tuple(size_t(0), seed), r);
```
Hmm AFAICT this only makes sense for `BigInt` (or similar types with arbitrary precision), for all other user-defined or subtyped types using `real` and the first overload makes more sense...

Also there's no need to save and compute the `length` is `hasLength` is available. Another lambda comes with a cost.

JackStouffer replied:
>Hmm AFAICT this only makes sense for BigInt (or similar types with arbitrary precision), for all other user-defined or subtyped types using real and the first overload makes more sense...

The problem is there's no way to know if the type doesn't use `alias this`. And if it does use `alias this`, then they can cast to `real` in order to get the other overload.

>Also there's no need to save and compute the length is hasLength is available. Another lambda comes with a cost.

ok

wilzbach replied:
> `The problem is there's no way to know if the type doesn't use alias this. And if it does use alias this, then they can cast to real in order to get the other overload.`

There's and that was my entire point: 

```d
import std.bigint;
import std.stdio;

void main(string[] args)
{
    struct MyFancyDouble
    {
        double v;
        alias v this;
    }
    pragma(msg, typeof(MyFancyDouble.init / size_t(0)));
    pragma(msg, typeof(BigInt.init / size_t(0)));
}
```

-> https://is.gd/rRbSaP

Or in other words: if the division by `size_t` yields a FloatingPoint type -> algorithm from overload 1 should be used.

JackStouffer commented on std/numeric.d:
```diff
@@ -3329,6 +3329,117 @@ C swapRealImag(C)(C input)
+The first overload of this function will return `T.init` if the range
+is empty. However, the second overload will return `seed` on empty ranges.
+
+This function is $(BIGOH r.length).
```
big O notations are adjectives, not nouns.

JackStouffer commented:
>AFAICT the second overload solely exists for arbitrary-precision types like BigInt and there aren't many precedent of math* or numeric caring about BigInt. gcd is the only one I know of.

The `seed` parameter on `sum` exists for the same purpose.

wilzbach commented:
> The seed parameter on sum exists for the same purpose.

Yet another reason for putting it into `std.algorithm.iteration` ;-)
Also the `std.algorithm.iteration.sum` of an empty array is always zero, no matter whether the element type is `int`, `double` or `BigInt`.

JackStouffer commented:
Ping @wilzbach @CyberShadow 

Moved it back to std.algorithm as that makes the most sense. Let's get a final decision on this and get it out of the queue.

CyberShadow commented:
> Ping @wilzbach @CyberShadow

Thanks!

I think the API is looking much better than the initial iteration. Still, math stuff is really not my area of expertise (I was off on a limb reviewing this in the first place), so deferring this to std.algorithm code owners.

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+Returns:
+    The mean of `r` when `r` is non-empty.
+*/
+T mean(T = real, R)(R r)
```
We should move away from the use of `real` as a default and treat it as a special interest type. Please use double here.

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+if (isInputRange!R &&
+    isNumeric!(ElementType!R) &&
+    !isInfinite!R &&
+    is(Unqual!(T) == T))
```
Why is the unqual needed?

andralex replied:
superfluous parens, use `Unqual!T`

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+
+    // Knuth & Welford mean calculation
+    // division per element is slower, but more accurate
+    foreach (e; r)
```
Avoid foreach, use explicit loops

JackStouffer replied:
Huh? We're moving away from `foreach`?

andralex replied:
it's a convenience engine and may copy stuff etc. Just use a loop and use r.front only once inside the function.

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+    // division per element is slower, but more accurate
+    foreach (e; r)
+    {
+        immutable T delta = (e - meanRes);
```
spurious parens

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+    foreach (e; r)
+    {
+        immutable T delta = (e - meanRes);
+        meanRes += delta / ++i;
```
better use `meanRes += delta / i++;` initializing `i` to 1

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+auto mean(R, T)(R r, T seed)
+if (isInputRange!R &&
+    !isNumeric!(ElementType!R) &&
+    is(typeof(r.front + seed)) &&
```
the type of + should be numeric

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+if (isInputRange!R &&
+    !isNumeric!(ElementType!R) &&
+    is(typeof(r.front + seed)) &&
+    is(typeof(r.front / size_t(1))) &&
```
should be numeric

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+
+        if (len > 0)
+            return sum(r, seed) / len;
+        else
```
superfluous

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4980,6 +4980,142 @@ private auto sumKahan(Result, R)(Result result, R r)
+    // types might be representing
+    static if (hasLength!R)
+    {
+        immutable len = r.length;
```
that's overdoing it - just use `r.length`

JackStouffer commented:
>the type of + should be numeric

In the BigInt case this isn't true, because the type of the addition is BigInt and `isNumeric!(BigInt) == false`

JackStouffer commented:
Thanks everyone for getting this through!
